changes/2025-09-12_introduce-metrics-interface/change.md

```diff
@@ -0,0 +1,169 @@
+
+## Definitions
+
+### Conventions used in this document
+
+The key words
+"MUST", "MUST NOT", "REQUIRED", "SHALL", "SHALL NOT",
+"SHOULD", "SHOULD NOT", "RECOMMENDED", "MAY", and "OPTIONAL"
+in this document are to be interpreted as described in
+[RFC 2119](https://tools.ietf.org/html/rfc2119).
+
+## Summary
+
+Existing users of Crypto Tools (CT) libraries do no have any insights as to
+how the librar(y/ies) behave(s) in their application.
+This can lead to frustrating debugging sessions where users
+are required to have explicit tests to assert they are using a particular feature
+correctly, or if customers are using any of the KMS keyrings users have to have
+AWS Cloudwatch open to verify their application is sending values users expect.
+This can be seen as a best practice and users may find this a good exercise;
+however, CT's libraries do not make debugging an easy task.
+
+A feature which allows customers to get real-time telemetry of their application's
+integration with CT's libraries would be welcomed by users and will deliver on the
+"easy to use and hard to misuse" tenet.
+
+Introducing a new interface that defines the operations that must be
+implemented in order to build a specification compliant MetricsAgent.
+
+## Requirements
+
+The interface should have three requirements.
+
+1. MUST be simple.
+1. MUST be extensible.
+
+The following is documented to signify its importance
+even though the interface is not able to make this guarantee.
+Every implementation of the proposed interface must
+ensure the following.
+
+1. MUST NOT block the application's execution thread.
+
+## Points of Integration
+
+To collect metrics across CT's library stack multiple points of integration
+are needed in order to collect metrics across CT's libraries.
+Generally, CT's libraries work as follows:
+
+_Note: Not every Client supports the Material Provider Library.
+The Diagram below assumes it to help the mental model._
+
+```mermaid
+sequenceDiagram
+    participant Client
+    box Material Provider Library
+      participant CMM
+      participant Keyring
+    end
+    loop Content Encryption
+        Client->>Client: Content Encryption
+    end
+    Client<<->>CMM: GetEncryption/Decryption Materials
+    CMM<<->>Keyring: OnEncrypt/OnDecrypt
+```
+
+To optionally emit metrics from a top level client,
+all customer facing APIs MUST be changed to optionally accept
+a Metrics Agent. This will allow customers to define and supply one top
+level Metrics Agent; this agent will get plumbed throughout CT's stack.
+
+For example, in the ESDK for Java this would look like:
+
+```java
+final AwsCrypto crypto = AwsCrypto.builder().build();
+// Create a Keyring
+final MaterialProviders matProv =
+  MaterialProviders.builder()
+      .MaterialProvidersConfig(MaterialProvidersConfig.builder().build())
+      .build();
+
+final IKeyring rawAesKeyring = matProv.CreateRawAesKeyring(keyringInput);
```

Comment:
Why are we implementing our own metric agent? I'm pretty sure there exists some framework which we can leverage. Customers can optionally provide the agent, or we do no-op (like logging frameworks such as slf4j etc).